### PR TITLE
feat: enhance test coverage for TransactionFactory (Access List & Address formatting) 

### LINKED ISSUE #5337: feat: enhance test coverage for TransactionFactory (Access List & Address formatting)
### Problem

formatAddress function used to format the address field in the access and authorization lists does not have its own dedicated unit tests focused solely on it.

### Solution

Add tests for formatAddress function.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by ManmathX**
### Description

This PR significantly improves the reliability of the pectra-upgrade branch by adding a comprehensive suite of unit tests for the newly introduced formatAccessList and formatAddress helpers in TransactionFactory.ts.

While the implementation for Pectra support was already present, these helpers were largely untested. This contribution fills that gap, ensuring that the normalization logic for access lists (required for EIP-2930, EIP-1559, and EIP-7702) is robust against raw data from the MirrorNode.

Expanded Test Coverage: I added 37 new unit tests to the transactionFactory.spec.ts, increasing the total from 24 to 61 tests. All tests are currently passing.
Improved Edge Case Handling: Covered multiple scenarios for formatAccessList and formatAddress, including handling of null/undefined inputs, short address padding, oversized address truncation, and correctly processing the EMPTY_HEX ("0x") string from MirrorNode.
Verified Propagation: Confirmed that the access_list is correctly propagated for EIP-2930, EIP-1559, and EIP-7702 transaction types.

### Related issue(s)

Fixes #5337  

### Testing Guide

1.
2.
3.

### Changes from original design (optional)

N/A

### Additional work needed (optional)

N/A

### Checklist

- [ ] I've assigned an assignee to this PR and related issue(s) (if applicable)
- [ ] I've assigned a label to this PR and related issue(s) (if applicable)
- [ ] I've assigned a milestone to this PR and related issue(s) (if applicable)
- [ ] I've updated documentation (code comments, README, etc. if applicable)
- [ ] I've done sufficient testing (unit, integration, etc.)

### CODE REVIEW

**Comment by ManmathX** [comment]
Hi @jasuwienas,

I hope you’re doing well!

I’ve been following the progress on the pectra-upgrade branch and wanted to contribute by strengthening the test infrastructure around the new transaction logic. I’ve added a comprehensive set of unit tests for the formatAccessList and formatAddress helpers in TransactionFactory.ts to ensure we’re handling all possible MirrorNode data scenarios robustly

**Review by silkshjasuwienas — DISMISSED** [review]
@ManmathX Thank you! Really great work. Thank you for your contribution once again. It is a really nice idea to include formatAddress function tests. But I've left few suggestions we should apply before merging it.

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 1]
```diff
@@ -719,4 +719,461 @@ describe('TransactionFactory', () => {
...
+     *
+     * This function routes the provided `input` through
+     * `createTransactionFromContractResult` using a mocked EIP-1559
+     * transaction payload (type = 2), and then extracts the resulting
+     * `accessList`.
+     *
+     * @param {any} input - Raw access list input.
+     * @returns {any[]} The normalized and sanitized access list.
+     */
+    const formatAccessList = (input: any): any[] =>
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: We already have some of these access list tests inclued, please remove the duplicates.

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 2]
```diff
@@ -719,4 +719,461 @@ describe('TransactionFactory', () => {
...
+
+    it('propagates access_list through createTransactionFromContractResult for type 1 (EIP-2930)', () => {
+      const result: any = createTransactionFromContractResult({
+        amount: 0,
+        from: '0x05fba803be258049a27b820088bab1cad2058871',
+        function_parameters: '0x',
+        gas_used: 21000,
+        gas_limit: 21000,
+        to: '0x0000000000000000000000000000000000000409',
+        hash: '0x' + 'aa'.repeat(32),
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: Let's use prepend0x function from formatter.ts.

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 3]
```diff
@@ -719,4 +719,461 @@ describe('TransactionFactory', () => {
...
+
+      expect(out).to.have.length(1);
+      expect(out[0].storageKeys).to.deep.eq([]);
+    });
+
+    it('propagates access_list through createTransactionFromContractResult for type 1 (EIP-2930)', () => {
+      const result: any = createTransactionFromContractResult({
+        amount: 0,
+        from: '0x05fba803be258049a27b820088bab1cad2058871',
+        function_parameters: '0x',
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: Let's use constants where possible (EMPTY_HEX here).

File: tests/server/acceptance/sendRawTransactionExtension.spec.ts [hunk 4]
```diff
@@ -240,6 +240,43 @@ describe('@sendRawTransactionExtension Acceptance Tests', function () {
...
+
+        // Now verify if this access list is present in the transaction fetched by eth_getBlockByNumber.
+        const block = await relay.call('eth_getBlockByNumber', [tx.blockNumber, true]);
+        expect(block).to.have.property('transactions').that.is.an('array');
+        const transactionInBlock = block.transactions.find((t: any) => t.hash === transactionHash);
+        expect(transactionInBlock).to.have.property('accessList').that.is.an('array');
+        expect(transactionInBlock.accessList).to.not.be.empty;
+      });
+
+      it('should fail when calling "eth_sendRawTransaction" with non-empty access list and access list not taken into consideration when calculating gas limit', async function () {
```
Comment (jasuwienas) [inline tests/server/acceptance/sendRawTransactionExtension.spec.ts:279]: If the codacy suggestion is easy to apply let's do it (we can use arrow function here).

**Review by jasuwienas — CHANGES_REQUESTED** [review]
@ManmathX Thank you! Really great work. Thank you for your contribution once again. It is a really nice idea to include formatAddress function tests. But I've left few suggestions we should apply before merging it .

**Comment by jasuwienas** [comment]
@ManmathX also make sure that all of your commits are correctly signed and verified, https://github.com/hiero-ledger/hiero-json-rpc-relay/pull/5315/checks?check_run_id=72481463943 , please :) .

**Comment by ManmathX** [comment]
Hi @jasuwienas — I’ve addressed all the review comments:

* Removed duplicate access list tests
* Replaced literals with `prepend0x` where applicable
* Used constants like `EMPTY_HEX`
* Applied the arrow function suggestion in the acceptance test

I also re-signed my commits so they’re now verified. Could you please take another look when you get a chance? Thanks!

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 5]
```diff
@@ -719,4 +722,467 @@ describe('TransactionFactory', () => {
       expect(out).to.have.property('extraField').equal('keep-me');
     });
   });
+
+  describe('formatAccessList', () => {
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: @ManmathX 
Previous section is already described as 'formatAccessList', what do you think about merging them into one? We don't need to have them separated.

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 6]
```diff
@@ -719,4 +722,467 @@ describe('TransactionFactory', () => {
...
+      const out = formatAccessList(undefined);
+      expect(out).to.deep.eq([]);
+    });
+
+    it('returns empty array for non-array input (string)', () => {
+      const out = formatAccessList('0xdeadbeef');
+      expect(out).to.deep.eq([]);
+    });
+
+    it('returns empty array for non-array input (number)', () => {
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: 123 was already tested here:
https://github.com/hiero-ledger/hiero-json-rpc-relay/blob/pectra-upgrade/tests/relay/lib/factories/transactionFactory.spec.ts#L403

We don't need to check the same thign 2 times in a row. Let's keep only 1 test for 123 and remove the other one  :) .

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 7]
```diff
@@ -719,4 +722,467 @@ describe('TransactionFactory', () => {
...
+
+      return result?.accessList ?? [];
+    };
+
+    it('returns empty array for null input', () => {
+      const out = formatAccessList(null);
+      expect(out).to.deep.eq([]);
+    });
+
+    it('returns empty array for undefined input', () => {
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: undefined was already tested here:
https://github.com/hiero-ledger/hiero-json-rpc-relay/blob/pectra-upgrade/tests/relay/lib/factories/transactionFactory.spec.ts#L403

We don't need to check the same thign 2 times in a row. Let's keep only 1 test for undefined and remove the other one :) .

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 8]
```diff
@@ -719,4 +722,467 @@ describe('TransactionFactory', () => {
...
+        type: 2,
+        v: 1,
+        nonce: 2,
+        access_list: input,
+      }) as Transaction1559 | null;
+
+      return result?.accessList ?? [];
+    };
+
+    it('returns empty array for null input', () => {
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: null was already tested here:
https://github.com/hiero-ledger/hiero-json-rpc-relay/blob/pectra-upgrade/tests/relay/lib/factories/transactionFactory.spec.ts#L403

We don't need to check the same thign 2 times in a row. Let's keep only 1 test for null and remove the other one  :) .

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 9]
```diff
@@ -719,4 +722,467 @@ describe('TransactionFactory', () => {
...
+      const out = formatAccessList(null);
+      expect(out).to.deep.eq([]);
+    });
+
+    it('returns empty array for undefined input', () => {
+      const out = formatAccessList(undefined);
+      expect(out).to.deep.eq([]);
+    });
+
+    it('returns empty array for non-array input (string)', () => {
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: We have whole set of tests for hex input (including malformed ones):

https://github.com/hiero-ledger/hiero-json-rpc-relay/blob/pectra-upgrade/tests/relay/lib/factories/transactionFactory.spec.ts#L403

We should place them there (but we also dont need duplicates), so I propose to skip this one test for now.

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 10]
```diff
@@ -719,4 +722,467 @@ describe('TransactionFactory', () => {
...
+      const out = formatAccessList('0xdeadbeef');
+      expect(out).to.deep.eq([]);
+    });
+
+    it('returns empty array for non-array input (number)', () => {
+      const out = formatAccessList(123);
+      expect(out).to.deep.eq([]);
+    });
+
+    it('returns empty array for empty array input', () => {
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: Wow, it's actually a great test, it's strange we don;t have it included yet :) .

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 11]
```diff
@@ -719,4 +722,467 @@ describe('TransactionFactory', () => {
...
+      const out = formatAccessList([]);
+      expect(out).to.deep.eq([]);
+    });
+
+    it('filters out null and non-object items from the array', () => {
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: We already have this test included here:

https://github.com/hiero-ledger/hiero-json-rpc-relay/blob/pectra-upgrade/tests/relay/lib/factories/transactionFactory.spec.ts#L403

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 12]
```diff
@@ -719,4 +722,467 @@ describe('TransactionFactory', () => {
...
+      const input = [['0x0000000000000000000000000000000000000001']];
+      const out = formatAccessList(input);
+
+      expect(out).to.have.length(1);
+      expect(out[0].address).to.equal(constants.ZERO_ADDRESS_HEX);
+      expect(out[0].storageKeys).to.deep.eq([]);
+    });
+
+    it('defaults storageKeys to empty array when explicitly set to undefined', () => {
+      const input = [{ address: '0x0000000000000000000000000000000000000409', storageKeys: undefined }];
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: We already ahve it tested here:
https://github.com/hiero-ledger/hiero-json-rpc-relay/blob/pectra-upgrade/tests/relay/lib/factories/transactionFactory.spec.ts#L462

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 13]
```diff
@@ -719,4 +722,467 @@ describe('TransactionFactory', () => {
...
+
+    it('defaults storageKeys to empty array when missing', () => {
+      const input = [{ address: '0x0000000000000000000000000000000000000409' }];
+      const out = formatAccessList(input);
+
+      expect(out).to.have.length(1);
+      expect(out[0].storageKeys).to.deep.eq([]);
+    });
+
+    it('defaults storageKeys to empty array when null', () => {
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: We already have it tested here: https://github.com/hiero-ledger/hiero-json-rpc-relay/blob/pectra-upgrade/tests/relay/lib/factories/transactionFactory.spec.ts#L462

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 14]
```diff
@@ -719,4 +722,467 @@ describe('TransactionFactory', () => {
...
+
+      expect(out[0].address).to.equal('0x0000000000000000000000000000000000001234');
+    });
+
+    it('normalizes addresses without 0x prefix', () => {
+      const input = [{ address: '0000000000000000000000000000000000000409', storageKeys: [] }];
+      const out = formatAccessList(input);
+
+      expect(out[0].address).to.equal('0x0000000000000000000000000000000000000409');
+    });
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: Address normalization is already tested here: https://github.com/hiero-ledger/hiero-json-rpc-relay/blob/pectra-upgrade/tests/relay/lib/factories/transactionFactory.spec.ts#L435

If we want to include new tests, let's remove the previous ones. (let's not check the same thing twice)

File: tests/relay/lib/factories/transactionFactory.spec.ts [hunk 15]
```diff
@@ -719,4 +722,467 @@ describe('TransactionFactory', () => {
...
+      const out = formatAccessList(input);
+
+      expect(out).to.have.length(2);
+      expect(out[0].address).to.equal('0x0000000000000000000000000000000000000001');
+      expect(out[0].storageKeys).to.deep.eq(['0xabc']);
+      expect(out[1].address).to.equal('0x0000000000000000000000000000000000000002');
+      expect(out[1].storageKeys).to.deep.eq(['0xdef', '0x123']);
+    });
+
+    it('does NOT preserve extra properties on access list items (unlike authorization list)', () => {
```
Comment (jasuwienas) [inline tests/relay/lib/factories/transactionFactory.spec.ts]: This test is great, but we should remove this one: https://github.com/hiero-ledger/hiero-json-rpc-relay/blob/pectra-upgrade/tests/relay/lib/factories/transactionFactory.spec.ts#L474 first in order  not to keep duplicates :) .

**Review by jasuwienas — COMMENTED** [review]
@ManmathX

Once again, really great job on this, there are a lot of excellent tests here that we should definitely include 👍 .

As I mentioned in the last review, a few of them overlap with the ones we already have and end up testing the same behavior. Could you take another pass to remove the duplicates? I’ve pointed out the ones I noticed! Feel free to decide which versions make the most sense to keep.

Thanks again for the solid work here!

**Comment by ManmathX** [comment]
Hi @jasuwienas, I’ve gone through your suggestions and made the necessary changes. Could you please review when you have a moment and let me know if everything looks good now or if any further improvements are needed? Thanks!

**Review by jasuwienas — APPROVED** [review]
LGTM!
Thanks a lot for the contribution and for adding the extra tests, they look solid and improve coverage nicely. Much appreciated!
